estimater.py

```diff
@@ -370,33 +377,40 @@ def track_one_among_noises(self, rgb, depth, K, iteration, pose_last, sample_num
         extra['pose_center'] = dpose
     return best_pose.detach().reshape(4,4)
 
-def sample_added_noise(pose, 
-                       current_pos_noise=0.02, 
-                       current_rot_noise=0.15,
-                       sample_n = 20):
-  if sample_n != 1:
-    #convert
-    org_pose = torch.tensor(pose)
-    org_pos, org_ori = matrix_to_pos_rotation_matrix(org_pose)
-    org_pos = org_pos.squeeze()
-    org_ori = org_ori.squeeze()
-    org_ori = matrix_to_quaternion(org_ori.unsqueeze(0)).squeeze()
-    pos = copy.deepcopy(org_pos).repeat(sample_n, 1)
-    ori = copy.deepcopy(org_ori).repeat(sample_n, 1)
-    pos_noise = torch.rand(size=(sample_n, 3)) * 2 -1 # sample from [-1, 1]
-    rot_noise = torch.rand(size=(sample_n, 4)) * 2 -1 # sample from [-1, 1]
-    # # ====================== Add noise ========================
-    position_noise = pos_noise * current_pos_noise
-    pos += position_noise          
-
-    # # add orientation loss
-    rotation_noise = rot_noise * current_rot_noise
-    ori += rotation_noise
-    ori = normalize_quaternion_torch(ori)
-    # =========================================================
-    ori = quaternion_to_matrix(ori)
-    pose = batch_pos_rot_matrix_to_matrix(pos, ori)
-    
-    org_pose = org_pose.squeeze()
-    pose = torch.cat((pose, org_pose.unsqueeze(0)), dim=0)
-  return torch.tensor(pose).reshape(-1, 4, 4)
+  def sample_added_noise(self,
+                         pose,
+                         depth,
+                         mask,
+                         K,
+                         sample_guess_translation,
+                         current_pos_noise=0.02, 
+                         current_rot_noise=0.15,
+                         sample_n = 20):
+    if sample_n != 1:
+      #convert
+      org_pose = torch.tensor(pose).unsqueeze(0)
+      org_pos, org_ori = matrix_to_pos_rotation_matrix(org_pose)
+      org_pos = org_pos.squeeze()
+      org_ori = org_ori.squeeze()
+      org_ori = matrix_to_quaternion(org_ori.unsqueeze(0)).squeeze()
+      pos = copy.deepcopy(org_pos).repeat(sample_n, 1)
+      ori = copy.deepcopy(org_ori).repeat(sample_n, 1)
+      pos_noise = torch.rand(size=(sample_n, 3)) * 2 -1 # sample from [-1, 1]
+      rot_noise = torch.rand(size=(sample_n, 4)) * 2 -1 # sample from [-1, 1]
+      # # ====================== Add noise ========================
+      if sample_guess_translation: # guess translation
+        pos = torch.tensor(self.guess_translation(depth=depth, mask=mask, K=K)).repeat(sample_n, 1)
+      else: # add noise on last timestep pose
+        position_noise = pos_noise * current_pos_noise
+        pos += position_noise          
```

Comment:
So the translation is always fixed and only rotation is sampled with noise? Any rationale for this?

Reply:
We modified the code to give noise at translation (See updated comment above). So now noise is added at both translation and rotation.